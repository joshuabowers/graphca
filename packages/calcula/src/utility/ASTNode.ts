import { Writer } from '../monads/writer'

export const $kind = Symbol('$kind')

export interface ASTNode {
  readonly [$kind]: string,
  readonly isPrimitive: boolean,
  readonly isUnary: boolean,
  readonly isBinary: boolean,
}

export type Kind<T extends string> = {readonly [$kind]: T}
export type Form<K extends string, R extends {}> = Kind<K> & R

export const areKindEqual = <T extends ASTNode, U extends ASTNode>(
  t: Writer<T>, u: Writer<U>
) => t.value[$kind] === u.value[$kind]
