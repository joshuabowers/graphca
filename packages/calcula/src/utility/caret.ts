import { Particle } from "./operation"
import { Unicode } from "../Unicode"

export enum CaretPosition {
  none,
  before,
  between,
  after
}

export function embedCaret(position: CaretPosition, argument: Particle[]): Particle[];
export function embedCaret(position: CaretPosition, left: Particle[], operator: string, right: Particle[]): Particle[];
export function embedCaret(position: CaretPosition, left: Particle[], operator?: string, right?: Particle[]): Particle[]{
  const result = [...left]
  if(position === CaretPosition.before){ result.unshift(Unicode.process) }
  else if(!right && position === CaretPosition.after){ 
    result.push(Unicode.process) 
  }
  if(operator && right){ 
    result.push(operator)
    if(position === CaretPosition.between){ result.push(Unicode.process) }
    result.push(right)
    if(position === CaretPosition.after){ result.push(Unicode.process) }
  }
  return result
}
