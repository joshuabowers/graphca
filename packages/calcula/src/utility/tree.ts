import { Writer } from "../monads/writer"

export enum Clades {
  primitive = 'Primitive',  // Constants: 10, 4 - 5i, true
  unary = 'Unary',          // Functions and Ops: not, abs, sin
  binary = 'Binary',        // Functions and Ops: and, log, polygamma
  variadic = 'Variadic'     // Variables: x, OhGodWhy
}

export enum Genera {
  arithmetic = 'Arithmetic',
  connective = 'Connective',
  inequalities = 'Inequalities',
  trigonometric = 'Trigonometric',
  arcus = 'Arcus',
  hyperbolic = 'Hyperbolic',
  areaHyperbolic = 'AreaHyperbolic',
  combinatorics = 'Combinatorics'
}

export enum Species {
  // Primitives
  real = 'Real',
  complex = 'Complex',
  boolean = 'Boolean',
  nil = 'Nil',
  nan = 'NaN',

  // Arithmetic
  add = 'Addition',
  multiply = 'Multiplication',
  raise = 'Exponentiation',

  // Functions
  abs = 'Absolute',

  // :: Inequalities
  equals = 'Equality',
  notEquals = 'Inequality',
  lessThan = 'LessThan',
  greaterThan = 'GreaterThan',
  lessThanEquals = 'LessThanEquals',
  greaterThanEquals = 'GreaterThanEquals',

  // :: Connectives
  not = 'Complement',
  and = 'Conjunction',
  or = 'Disjunction',
  xor = 'ExclusiveDisjunction',
  implies = 'Implication',
  nand = 'AlternativeDenial',
  nor = 'JointDenial',
  xnor = 'Biconditional',
  converse = 'ConverseImplication',

  // :: Trigonometric
  sin = 'Sine',
  cos = 'Cosine',
  tan = 'Tangent',
  csc = 'Cosecant',
  sec = 'Secant',
  cot = 'Cotangent',

  // Variables
  variable = 'Variable'
}

export type TreeNode = {
  readonly clade: Clades,
  readonly genus?: Genera,
  readonly species: Species
}

export type DerivedNode<T extends ((...args: any[]) => any)> = 
  ReturnType<T> extends Writer<infer U> ? U : never

export type TreeNodeGuardFn<T extends TreeNode> = 
  (value: Writer<TreeNode>) => value is Writer<T>

// values should be Writer<TreeNode>, narrowed to Writer<T>
export const isClade = <T extends TreeNode>(clade: Clades) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.clade === clade

export const isGenus = <T extends TreeNode>(genus: Genera) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.genus === genus

export const isSpecies = <T extends TreeNode>(species: Species) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.species === species

export const notAny = (...args: Species[]) => {
  const exclude = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && !exclude.has(t.value.species)
}

export const any = (...args: Species[]) => {
  const include = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && include.has(t.value.species)
}
