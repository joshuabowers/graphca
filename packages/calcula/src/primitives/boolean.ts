import { Species, isSpecies } from '../utility/tree'
import { Boolean, primitive } from '../closures/primitive'
import { isBoolean as isSystemBoolean } from '../utility/valuePredicates'
export { Boolean }

export const boolean = primitive<boolean, {value: boolean}, Boolean>(
  isSystemBoolean,
  value => ({value}),
  Species.boolean
)(
  create => r => [create(r.value !== 0), 'cast to boolean'],
  create => c => [create(c.a !== 0 || c.b !== 0), 'cast to boolean'],
  _create => b => [b, '']
)()

export const isBoolean = isSpecies<Boolean>(Species.boolean)
