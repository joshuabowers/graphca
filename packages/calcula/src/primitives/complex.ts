import { Species } from "../utility/tree"
import { Complex, primitive } from "../closures/primitive"
import { isNumberTuple } from "../utility/valuePredicates"
export { Complex }

export const [complex, isComplex, $complex] = 
  primitive<[number, number], {a: number, b: number}, Complex>(
  isNumberTuple,
  ([a, b]) => ({a, b}),
  Species.complex
)(
  r => [r.value, 0],
  c => [c.a, c.b],
  b => [b.value ? 1 : 0, 0]
)()

export const ComplexInfinity = complex([Infinity, NaN])
