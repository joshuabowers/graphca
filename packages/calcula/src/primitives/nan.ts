import { Writer, unit } from '../monads/writer'
import { Species, Clades, isSpecies } from '../utility/tree'
// import { $kind } from '../utility/ASTNode'
import { NaN } from "../closures/primitive"
export { NaN }

export const nan: Writer<NaN> = unit({
  clade: Clades.primitive,
  species: Species.nan,
  value: NaN
})

export const isNaN = isSpecies<NaN>(Species.nan)
