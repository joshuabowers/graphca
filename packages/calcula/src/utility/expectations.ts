import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
