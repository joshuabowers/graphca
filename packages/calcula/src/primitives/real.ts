import { Species } from '../utility/tree'
import { Real, primitive } from '../closures/primitive'
import { isNumericOrString } from '../utility/valuePredicates'
import { numeric } from '../utility/numeric'
import { nan } from './nan'
export { Real }

export const [real, isReal, $real] = 
  primitive<[number|string], number, Real>(
  isNumericOrString,
  value => Number(value),
  Species.real,
  r => numeric(r.raw)
)(
  r => [r.raw],
  c => [c.raw.a],
  b => [b.raw ? 1 : 0]
)(
  when => [
    when(v => Number.isNaN(v), [nan, 'incalculable'])
  ]
)

export const EulerMascheroni = real(0.57721566490153286060)
