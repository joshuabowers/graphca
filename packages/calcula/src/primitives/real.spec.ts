import { Clades, Species } from '../utility/tree'
import { real, $real } from './real'
import { boolean } from './boolean'
import { complex } from './complex'
import { nan } from './nan'
import { expectToEqualWithSnapshot } from '../utility/expectations'

describe('$real', () => {
  it('returns a Real given a numeric parameter', () => {
    expect($real(5)).toEqual({
      clade: Clades.primitive, genus: undefined, species: Species.real, 
      raw: 5
    })
  })
})

describe('real', () => {
  it('returns a Writer<Real> for a number input', () => {
    expectToEqualWithSnapshot(
      real(5), $real(5)
    )
  })

  it('returns a Writer<Real> for a real input', () => {
    expectToEqualWithSnapshot(
      real(real(5)), $real(5)
    )
  })

  it('returns a Writer<Real> for a complex input', () => {
    expectToEqualWithSnapshot(
      real(complex(1, 2)), $real(1)
    )
  })

  it('returns a Writer<Real> for a boolean input', () => {
    expectToEqualWithSnapshot(
      real(boolean(true)), $real(1)
    )
  })

  describe('of special values', () => {
    it('correctly logs e', () => {
      expectToEqualWithSnapshot(
        real(Math.E), $real(Math.E)
      )
    })

    it('correctly logs pi', () => {
      expectToEqualWithSnapshot(
        real(Math.PI), $real(Math.PI)
      )
    })

    it('correctly logs positive infinity', () => {
      expectToEqualWithSnapshot(
        real(Infinity), $real(Infinity)
      )
    })

    it('correctly logs negative infinity', () => {
      expectToEqualWithSnapshot(
        real(-Infinity), $real(-Infinity)
      )
    })

    it('rewrites numeric NaN with Writer<NaN>', () => {
      expectToEqualWithSnapshot(
        real(NaN), nan
      )
    })
  })
})
