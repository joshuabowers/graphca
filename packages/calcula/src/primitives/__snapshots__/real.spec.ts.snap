// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`real of special values correctly logs e 1`] = `
[
  [
    "𝒆",
    "created real",
  ],
]
`;

exports[`real of special values correctly logs negative infinity 1`] = `
[
  [
    "-∞",
    "created real",
  ],
]
`;

exports[`real of special values correctly logs pi 1`] = `
[
  [
    "𝜋",
    "created real",
  ],
]
`;

exports[`real of special values correctly logs positive infinity 1`] = `
[
  [
    "∞",
    "created real",
  ],
]
`;

exports[`real of special values rewrites numeric NaN with Writer<NaN> 1`] = `
[
  [
    "NaN",
    "incalculable",
  ],
  [
    "NaN",
    "not a number",
  ],
]
`;

exports[`real returns a Writer<Real> for a boolean input 1`] = `
[
  [
    "true",
    "created boolean",
  ],
  [
    "1",
    "cast to Real from Boolean",
  ],
]
`;

exports[`real returns a Writer<Real> for a complex input 1`] = `
[
  [
    "1+2𝒊",
    "created complex",
  ],
  [
    "1",
    "cast to Real from Complex",
  ],
]
`;

exports[`real returns a Writer<Real> for a number input 1`] = `
[
  [
    "5",
    "created real",
  ],
]
`;

exports[`real returns a Writer<Real> for a real input 1`] = `
[
  [
    "5",
    "created real",
  ],
  [
    "5",
    "copied Real",
  ],
]
`;
